nncf/experimental/common/tensor_statistics/collectors.py

```diff
@@ -39,6 +39,7 @@ def __init__(self, reduction_shape: Optional[ReductionShape] = None, inplace: bo
 
         """
         self._reduction_shape = reduction_shape
+        self._init_reduction_shape = reduction_shape
```

Comment:
The problem is that reduction_shape is needed to calculate reducer hash value to distinguish two different reducer with equal class. In case reduction shape is None at the begining it's calculated on first iteration
https://github.com/openvinotoolkit/nncf/blob/16d18e895bcb45a7a40bf36e2961c9ae70495222/nncf/experimental/common/tensor_statistics/collectors.py#L94-L95 so, reducer hash during registration and reducer shape after first iteration are different, so maps inside `TensorCollector` will be invalid after the first iteration. This `_init_reduction_shape` doesn't change during reducer life so it's safe for hash calculation

Reply:
It looks like if reduction_shape is None than the reducer should reduce by all dimensions. I suggest to implement this logic in each reducer not in the base class. You can implement the method to create reduction shape by tensor and use it in any inhered class.

Reply:
Done, please check